Drop Ruby < 2.3 support

pirj opened the pull request:
Better viewed as https://github.com/rspec/rspec-expectations/pull/1231/files?diff=split&w=1


Sibling PRs:
 - https://github.com/rspec/rspec-support/pull/436
 - https://github.com/rspec/rspec-mocks/pull/1349
 - https://github.com/rspec/rspec-core/pull/2787

pirj commented on maintenance-branch:
```diff
@@ -1 +1 @@
-4-0-dev
+drop-old-rubies
```
⚠️ This change is temporary and needs to be removed before merging, or soon after ⚠️

JonRowe replied:
This PR should be capable of running off 4-0-dev? What makes it dependant?

pirj replied:
Reverted 👍

JonRowe commented on Gemfile:
```diff
@@ -37,50 +29,14 @@ end
-if RUBY_VERSION < '1.9.2'
-  gem 'contracts', '~> 0.15.0' # is a dependency of aruba
-end
+gem 'ffi', '> 1.9.24' # prevent Github security vulnerability warning
```
Shouldn't this be `~> 1.12.0`? `1.13` requires Rubygems and we dont.

JonRowe commented on lib/rspec/expectations/failure_aggregator.rb:
```diff
@@ -52,7 +52,7 @@ def call(failure, options)
     private
 
-      if RSpec::Support::Ruby.jruby? && RSpec::Support::Ruby.jruby_version < '9.2.0.0'
+      if RSpec::Support::Ruby.jruby? && JRUBY_VERSION < '9.2.0.0'
```
Revert this, we're keeping the original comparable check because it works correctly with the JRuby strings.

JonRowe commented on lib/rspec/matchers/built_in/has.rb:
```diff
@@ -80,11 +71,7 @@ def predicate_matches?(value=true)
-          # work around this bug. For an example of this bug, see:
-          # https://travis-ci.org/rspec/rspec-expectations/jobs/27549635
-          self.class::REGEX.match(@method_name.to_s).captures.first
+          self.class::REGEX.match(@method_name).captures.first
```
This comment implies there was a better historical implementation, presumedly using a string rather than a regex, if you're going to change the implementation in a non compatible way might as well revert to that?

pirj replied:
I didn't see a good reason for this change:
```diff
- self.class::REGEX.match(@method_name).captures.first
+ @method_name.match(self.class::REGEX).captures.first
```

The reason for dropping `to_s` is that Regex#match works fine with a Symbol argument, but apparently it didn't before 2.0.
Also, Symbol#match works, so we actually can use (revert to some original) `@method_name.match` if we decide to.

I don't have any preference here. Just took the option with less churn.

JonRowe replied:
Can we benchmark the various options and see which is best?

pirj replied:
```ruby
require 'benchmark/ips'

BE_PREDICATE_REGEX = /^(?:be_(?:an?_)?)(.*)/
HAS_REGEX = /^(?:have_)(.*)/

BE_METHOD_NAMES = [:be_an_apple, :be_apple]
HAS_METHOD_NAMES = [:have_apples, :have_apple]

# Sanity checks
BE_METHOD_NAMES.each do |method_name|
  fail unless BE_PREDICATE_REGEX.match(method_name).captures == method_name.match(BE_PREDICATE_REGEX).captures
end

HAS_METHOD_NAMES.each do |method_name|
  fail unless HAS_REGEX.match(method_name).captures == method_name.match(HAS_REGEX).captures
end

report = Benchmark.ips do |x|
  x.report('Regex#match') do
    BE_METHOD_NAMES.each do |method_name|
      BE_PREDICATE_REGEX.match(method_name).captures.first
    end
    HAS_METHOD_NAMES.each do |method_name|
      HAS_REGEX.match(method_name).captures.first
    end
  end

  x.report('Regex#match(to_s)') do
    BE_METHOD_NAMES.each do |method_name|
      BE_PREDICATE_REGEX.match(method_name.to_s).captures.first
    end
    HAS_METHOD_NAMES.each do |method_name|
      HAS_REGEX.match(method_name.to_s).captures.first
    end
  end

  x.report('Symbol#match') do
    BE_METHOD_NAMES.each do |method_name|
      method_name.match(BE_PREDICATE_REGEX).captures.first
    end
    HAS_METHOD_NAMES.each do |method_name|
      method_name.match(HAS_REGEX).captures.first
    end
  end
end

puts report
```

```
Calculating -------------------------------------
         Regex#match    154.233k (± 7.0%) i/s -    768.576k in   5.010355s
   Regex#match(to_s)    131.795k (± 7.5%) i/s -    657.359k in   5.019755s
        Symbol#match    148.697k (± 7.6%) i/s -    743.634k in   5.034812s
```
Consistent across runs, `Regex#match` is the winner.

JonRowe commented on lib/rspec/matchers/built_in/match.rb:
```diff
@@ -67,27 +67,17 @@ def match_captures(expected, actual)
-      # Used to wrap match data and make it reliable for 1.8.7
-      class ReliableMatchData
+      # @private
+      class MatchData
```
Given the comment, you should be able to remove this entire class and use match_data directly now, otherwise revert.

pirj replied:
Aha, nice catch!

JonRowe commented on spec/rspec/expectations/failure_aggregator_spec.rb:
```diff
@@ -67,7 +67,7 @@ def common_contiguous_frame_percent(failure, aggregate)
     it 'ensures the sub-failure backtraces are in a form that overlaps with the aggregated failure backtrace' do
       if RSpec::Support::Ruby.jruby?
-        pending "This is broken on 9.2.x.x" unless RSpec::Support::Ruby.jruby_version < '9.2.0.0'
+        pending "This is broken on 9.2.x.x" unless JRUBY_VERSION < '9.2.0.0'
```
Revert.

JonRowe reviewed (changes requested):
Good work @pirj, just a few things to put back and a few questions.

pirj commented:
Updates:
 - reverted `jruby_version` usages
 - pinned `ffi` to `~> 1.12.0`
 - added Changelog entry under "Breaking changes"
 - removed `MatchData` wrapper

JonRowe commented on Changelog.md:
```diff
@@ -1,6 +1,10 @@
-### Development
+# Development
```
```suggestion
### Development
```

JonRowe commented on Changelog.md:
```diff
@@ -1,6 +1,10 @@
+# Development
 [Full Changelog](http://github.com/rspec/rspec-expectations/compare/v3.10.0...main)
 
+## Breaking Changes
```
```suggestion
Breaking Changes:
```

JonRowe replied:
Sorry I didn't mean to break consistency with my first suggestion

pirj replied:
No worries!

JonRowe commented on cucumber.yml:
```diff
@@ -1,10 +1,2 @@
-%>
-
-default: --require features --tags <%= tags("~@wip") %> --format progress
+default: --require features --format progress
```
```suggestion
default: --require features --tags ~@wip --format progress
```

JonRowe commented on lib/rspec/matchers/built_in/count_expectation.rb:
```diff
@@ -56,14 +56,8 @@ def times
-            number >= count.first && number <= count.last
-          end
+        def cover?(count, number)
+          count.cover?(number)
```
Any reason why this wasn't refactored to use it directly?

pirj replied:
Nice catch. Let me inline this quickly.

JonRowe reviewed (changes requested):
Spotted a couple more things

pirj commented:
Restored the branch, `rspec-support` depends on it.
